LibraBFT/Impl/Consensus/SafetyRules/Properties/SafetyRules.agda

```diff
@@ -124,14 +137,77 @@ module verifyAndUpdateLastVoteRoundMSpec (round : Round) (safetyData : SafetyDat
   proj₂ (contract Post pre pfOk pfBail) ¬r>lvr = pfBail (toWitnessF ¬r>lvr)
 
 
-module verifyQcMSpec (qc : QuorumCert) where
-  postulate
-    -- TODO-1: This contract needs refinement.
-    contract
-      : ∀ P pre
-        → (P (inj₁ fakeErr) pre [])
-        → (P (inj₂ unit) pre [])
-        → LBFT-weakestPre (verifyQcM qc) P pre
+module verifyQcMSpec (self : QuorumCert) where
+  getVv : RoundManager → ValidatorVerifier
+  getVv = _^∙ (rmEpochState ∙ esVerifier)
+
+  -- See comment on contract below to understand the motivation for stating and proving the property
+  -- this way.
+
+  P' : RoundManager → RWST-Post Output RoundManager (Either ErrLog Unit)
```

Comment:
Perhaps we can rename this to `Contract` (and replace the pattern-matching lambda with two defining clauses, just to clean it up a bit?)

Reply:
Good idea, done, thanks.